File: app/src/main/java/com/futabooo/android/booklife/screen/search/SearchActivity.java
package com.futabooo.android.booklife.screen.search;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.SearchView;
import android.view.Menu;
import android.view.MenuItem;
import com.futabooo.android.booklife.BookLife;
import com.futabooo.android.booklife.R;
import com.futabooo.android.booklife.databinding.ActivitySearchBinding;
import com.futabooo.android.booklife.model.Book;
import com.futabooo.android.booklife.model.SearchResultResource;
import com.futabooo.android.booklife.screen.BookListMenu;
import com.futabooo.android.booklife.screen.addbook.AddBookDialogFragment;
import com.futabooo.android.booklife.screen.bookdetail.BookDetailActivity;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import javax.inject.Inject;
import okhttp3.ResponseBody;
import org.jsoup.Jsoup;
import retrofit2.Retrofit;
import timber.log.Timber;

public class SearchActivity extends AppCompatActivity
    implements BookRegisterBottomSheetDialogFragment.OnBottomSheetActionListener,
    AddBookDialogFragment.OnAddBookActionListener {

  @Inject Retrofit retrofit;
  @Inject SharedPreferences sharedPreferences;

  private String userId;
  private String csrfToken;

  private ActivitySearchBinding binding;

  private SearchResultAdapter resultAdapter;

  public static Intent createIntent(Context context) {
    return new Intent(context, SearchActivity.class);
  }

  @Override protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    ((BookLife) getApplication()).getNetComponent().inject(this);
    userId = sharedPreferences.getString("user_id","");

    binding = DataBindingUtil.setContentView(this, R.layout.activity_search);
    setSupportActionBar(binding.activitySearchToolbar);
    getSupportActionBar().setDisplayShowTitleEnabled(false);
    getSupportActionBar().setDisplayHomeAsUpEnabled(true);

    binding.activitySearchResultList.setLayoutManager(new LinearLayoutManager(SearchActivity.this));
  }

  @Override public boolean onCreateOptionsMenu(Menu menu) {
    binding.activitySearchToolbar.inflateMenu(R.menu.activity_search_menu);
    final SearchView searchView =
        (SearchView) binding.activitySearchToolbar.getMenu().findItem(R.id.menu_search).getActionView();
    searchView.setIconified(false);
    searchView.setQueryHint(getString(R.string.search_hint));
    searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
      @Override public boolean onQueryTextSubmit(String query){
        searchView.clearFocus();
        searchBooks(query);
        return false;
      }

      @Override public boolean onQueryTextChange(String newText) {
        return false;
      }
    });
    return super.onCreateOptionsMenu(menu);
  }

  @Override public boolean onOptionsItemSelected(MenuItem item) {
    int id = item.getItemId();

    switch (id) {
      case android.R.id.home:
        finish();
        return true;
    }
    return super.onOptionsItemSelected(item);
  }

  private void searchBooks(final String keyword) {
    Observable<ResponseBody> observable = retrofit.create(SearchService.class).get(keyword);
    observable.subscribeOn(Schedulers.io()).flatMap(new Function<ResponseBody, ObservableSource<JsonObject>>() {
      @Override public ObservableSource<JsonObject> apply(ResponseBody responseBody) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(responseBody.byteStream()));
        StringBuffer result = null;
        try {
          result = new StringBuffer();
          String line;
          while ((line = reader.readLine()) != null) {
            result.append(line);
          }
        } catch (IOException e) {
          Timber.e(e);
        }

        csrfToken = Jsoup.parse(result.toString()).select("meta[name=csrf-token]").get(0).attr("content");
        return retrofit.create(SearchService.class).getJson(csrfToken, keyword, "recommended", "japanese", 0, 20);
      }
    }).observeOn(AndroidSchedulers.mainThread()).subscribe(new Observer<JsonObject>() {
      @Override public void onSubscribe(Disposable d) {
      }

      @Override public void onNext(JsonObject value) {
        JsonArray jsonArray = value.getAsJsonArray("resources");
        Gson gson = new Gson();
        SearchResultResource[] resources = gson.fromJson(jsonArray, SearchResultResource[].class);
        resultAdapter = new SearchResultAdapter(SearchActivity.this, resources);
        binding.activitySearchResultList.setAdapter(resultAdapter);
        resultAdapter.setOnCardClickListener(new SearchResultAdapter.OnCardClickListener() {
          @Override public void onCardClick(SearchResultAdapter adapter, int position, Book book) {
            Intent intent = BookDetailActivity.createIntent(SearchActivity.this, book.getId());
            startActivity(intent);
          }

          @Override public void onRegisterClick(int bookId) {
            BookRegisterBottomSheetDialogFragment dialogFragment =
                BookRegisterBottomSheetDialogFragment.newInstance(bookId);
            dialogFragment.show(getSupportFragmentManager(), "bottom_sheet");
          }
        });
      }

      @Override public void onError(Throwable e) {
        Timber.e(e);
      }

      @Override public void onComplete() {
      }
    });
  }

  @Override public void onBottomSheetAction(BookListMenu bookListMenu, int bookId) {
    switch (bookListMenu) {
      case READ:
        //startActivity(BookEditActivity.createIntent(SearchActivity.this, asin));
        AddBookDialogFragment dialogFragment = AddBookDialogFragment.newInstance(csrfToken, userId, bookId);
        dialogFragment.show(getSupportFragmentManager(), "add_book_dialog");
        break;
      case READING:
      case TO_READ:
      case QUITTED:
        Observable<JsonObject> observable =
            retrofit.create(ActionService.class).addBook(csrfToken, userId, bookListMenu.getKey(), bookId);
        observable.subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(new Observer<JsonObject>() {
              @Override public void onSubscribe(Disposable d) {
              }

              @Override public void onNext(JsonObject value) {
              }

              @Override public void onError(Throwable e) {
                Timber.e(e);
              }

              @Override public void onComplete() {
              }
            });
    }
  }

  @Override public void onRegister(int bookId) {
    Snackbar.make(findViewById(android.R.id.content), getString(R.string.book_registered, String.valueOf(bookId)),
        Snackbar.LENGTH_LONG).show();
  }
}

File: app/src/main/kotlin/com/futabooo/android/booklife/screen/search/SearchActivity.kt
package com.futabooo.android.booklife.screen.search

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.SearchView
import android.view.Menu
import android.view.MenuItem
import com.futabooo.android.booklife.BookLife
import com.futabooo.android.booklife.InfiniteScrollListener
import com.futabooo.android.booklife.R
import com.futabooo.android.booklife.databinding.ActivitySearchBinding
import com.futabooo.android.booklife.extensions.observeOnUI
import com.futabooo.android.booklife.extensions.subscribeOnIO
import com.futabooo.android.booklife.model.SearchResultResource
import com.futabooo.android.booklife.screen.BookListMenu
import com.futabooo.android.booklife.screen.addbook.AddBookDialogFragment
import com.futabooo.android.booklife.screen.bookdetail.BookDetailActivity
import com.google.gson.Gson
import io.reactivex.rxkotlin.subscribeBy
import org.jsoup.Jsoup
import retrofit2.Retrofit
import timber.log.Timber
import java.io.BufferedReader
import java.io.InputStreamReader
import javax.inject.Inject

class SearchActivity : AppCompatActivity(), BookRegisterBottomSheetDialogFragment.OnBottomSheetActionListener, AddBookDialogFragment.OnAddBookActionListener {

  @Inject lateinit var retrofit: Retrofit
  @Inject lateinit var sharedPreferences: SharedPreferences

  private lateinit var binding: ActivitySearchBinding
  private lateinit var resultAdapter: SearchResultAdapter

  val userId by lazy { sharedPreferences.getInt("user_id", 0) }
  private lateinit var csrfToken: String
  private lateinit var keyword: String
  val limit: Int = 20
  var offset: Int = 0

  companion object {
    private val EXTRA_ISBN = "isbn"

    fun createIntent(context: Context, isbn: String = "") =
        Intent(context, SearchActivity::class.java).apply {
          putExtra(EXTRA_ISBN, isbn)
        }
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    (application as BookLife).netComponent.inject(this)

    binding = DataBindingUtil.setContentView<ActivitySearchBinding>(this, R.layout.activity_search).apply {
      setSupportActionBar(activitySearchToolbar)
      supportActionBar?.setDisplayShowTitleEnabled(false)
      supportActionBar?.setDisplayHomeAsUpEnabled(true)

      val layoutManager = LinearLayoutManager(this@SearchActivity)
      activitySearchResultList.layoutManager = layoutManager
      resultAdapter = SearchResultAdapter(mutableListOf(),
          { view, bookId ->
            when (view.id) {
              R.id.search_result_item -> {
                startActivity(BookDetailActivity.createIntent(this@SearchActivity, bookId))
              }
              R.id.search_result_book_action -> {
                BookRegisterBottomSheetDialogFragment.newInstance(bookId).show(supportFragmentManager, "bottom_sheet")
              }
            }
          })
      activitySearchResultList.adapter = resultAdapter
      activitySearchResultList.addOnScrollListener(InfiniteScrollListener({ searchBooks(keyword) }, layoutManager))
    }

    intent.getStringExtra(EXTRA_ISBN).let { searchBooks(it) }
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    binding.activitySearchToolbar.inflateMenu(R.menu.activity_search_menu)
    binding.activitySearchToolbar.menu.findItem(R.id.menu_search).actionView.also {
      it as SearchView
      it.isIconified = false
      it.queryHint = getString(R.string.search_hint)
      it.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
        override fun onQueryTextSubmit(query: String): Boolean {
          it.clearFocus()
          keyword = query
          searchBooks(keyword)
          return false
        }

        override fun onQueryTextChange(newText: String): Boolean {
          return false
        }
      })
    }
    return super.onCreateOptionsMenu(menu)
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    val id = item.itemId

    when (id) {
      android.R.id.home -> {
        finish()
        return true
      }
    }
    return super.onOptionsItemSelected(item)
  }

  private fun searchBooks(keyword: String) {
    retrofit.create(SearchService::class.java).get(keyword)
        .flatMap {
          val reader = BufferedReader(InputStreamReader(it.byteStream()))
          val result = reader.readLines().filter(String::isNotBlank).toList()
          csrfToken = Jsoup.parse(result.toString()).select("meta[name=csrf-token]")[0].attr("content")
          retrofit.create(SearchService::class.java).getJson(csrfToken, keyword, "recommended", "japanese", offset,
              limit)
        }
        .subscribeOnIO
        .observeOnUI
        .doOnSubscribe { resultAdapter.showProgress(true) }
        .doFinally { resultAdapter.showProgress(false) }
        .subscribeBy(
            onNext = {
              val jsonArray = it.getAsJsonArray("resources")
              val resources = Gson().fromJson(jsonArray, Array<SearchResultResource>::class.java)
              resultAdapter.addAll(resources.toMutableList())
              offset += limit
            },
            onError = { Timber.e(it, it.message) }
        )
  }

  override fun onBottomSheetAction(bookListMenu: BookListMenu, bookId: Int) {
    when (bookListMenu) {
      BookListMenu.READ -> {
        AddBookDialogFragment.newInstance(csrfToken, userId, bookId).show(supportFragmentManager, "add_book_dialog")
      }
      BookListMenu.READING, BookListMenu.TO_READ, BookListMenu.QUITTED -> {
        retrofit.create(ActionService::class.java).addBook(csrfToken, userId, bookListMenu.key, bookId)
            .subscribeOnIO
            .observeOnUI
            .subscribeBy(
                onError = { Timber.e(it, it.message) }
            )
      }
    }
  }

  override fun onRegister(bookId: Int) {
    Snackbar.make(findViewById(android.R.id.content), getString(R.string.book_registered, bookId.toString()),
        Snackbar.LENGTH_LONG).show()
  }

  override fun onUpdate(bookId: Int) {
    Snackbar.make(findViewById(android.R.id.content), getString(R.string.book_update, bookId.toString()),
        Snackbar.LENGTH_LONG).show()
  }

  override fun onError() {
    Snackbar.make(findViewById(android.R.id.content), getString(R.string.error_review_update),
        Snackbar.LENGTH_LONG).show()
  }
}
